kore-rpc-types/src/Kore/Syntax/Json/Types.hs

```diff
@@ -0,0 +1,344 @@
+      KJNext
+        { sort :: Sort
+        , dest :: KorePattern
+        }
+    | -- | source rewrites to dest (same sort)
+      KJRewrites
+        { sort :: Sort
+        , source :: KorePattern
+        , dest :: KorePattern
+        }
+    | -- | domain value, a string literal with a sort
+      KJDV
+        { sort :: Sort
+        , value :: Text
+        }
+    | -- syntactic sugar
+
+      -- | left/right associative or-cascade
+      KJMultiOr
+        { assoc :: LeftRight
+        , sort :: Sort
+        , argss :: NE.NonEmpty KorePattern
+        }
+    | -- TODO textual parser also understands And/Implies/Iff
+
+      -- | left/right associative app pattern
+      KJMultiApp
+        { assoc :: LeftRight
+        , symbol :: Id -- may start by a '\\'
+        , sorts :: [Sort]
+        , argss :: NE.NonEmpty KorePattern
+        }
+    deriving stock (Eq, Show, Generic)
+
+instance ToJSON KorePattern where
+    toJSON = genericToJSON codecOptions
+
+instance FromJSON KorePattern where
+    parseJSON v = genericParseJSON codecOptions v >>= lexicalCheck
+
+codecOptions :: Json.Options
+codecOptions =
+    Json.defaultOptions
+        { constructorTagModifier
+        , omitNothingFields = True
+        , sumEncoding = TaggedObject "tag" "contents"
+        , unwrapUnaryRecords = True
+        , tagSingleConstructors = True
+        , rejectUnknownFields = True
+        }
+  where
+    constructorTagModifier = \case
+        'K' : 'J' : rest -> rest
+        other -> other
+
+----------------------------------------
+-- Identifiers and lexical checks
+
+newtype Id = Id {getId :: Text}
+    deriving stock (Eq, Show, Ord, Generic)
+    deriving newtype (ToJSON, FromJSON)
+
+{- | Performs a (shallow, top-level, no recursion) lexical check of
+ identifiers contained in the given node. For details see the check
+ functions below.
+-}
+lexicalCheck :: KorePattern -> Json.Parser KorePattern
+lexicalCheck p =
+    case p of
+        KJEVar{name = Id n} ->
+            reportErrors n "element variable" checkIdChars
+        KJSVar{name = Id n} ->
+            reportErrors n "set variable" checkSVarName
+        KJApp{name = Id n} ->
+            reportErrors n "app symbol" checkSymbolName
+        KJForall{var = Id name} ->
+            reportErrors name "quantifier variable" checkIdChars
+        KJExists{var = Id name} ->
+            reportErrors name "quantifier variable" checkIdChars
+        KJMu{var = Id name} ->
+            reportErrors name "fixpoint expression variable" checkSVarName
+        KJNu{var = Id name} ->
+            reportErrors name "fixpoint expression variable" checkSVarName
+        -- KJDV{value = txt} ->
+        --     reportErrors txt "domain value string" checkStringChars
+        -- KJString txt ->
+        --     reportErrors txt "string literal" checkStringChars
+        -- Input supports std Unicode (as per json spec). toJSON could
+        -- check that only allowed escape sequences will be generated.
```

Comment:
not sure what we want to do here as this is the point the old and new back-end differ... @ana-pantilie @jberthold?

Reply:
I would say restrict to `latin1` as we are doing in the new back-end. 
We should probably investigate the old code path to find out why the byte arrays aren't a problem here but we need to use the version with restrictions in the new back-end until then.